### PR TITLE
Adding new automated prefab tests for adding an entity

### PR DESCRIPTION
**Pull request by dm-amzn**
Signed-off-by: dm-amzn <7991771+dm-amzn@users.noreply.github.com>

## What does this PR do?

Adding new automated prefab tests for adding an entity.

_Please add links to any issues, RFCs or other items that are relevant to this PR._

## How was this PR tested?

Ran new test locally to verify success. Ran periodic_test_profile on internal Jenkins instance to verify success.

### CODE REVIEW

**Review by forkercat — COMMENTED** [review]
looks like the test files are empty. Is it a git issue?

File: AutomatedTesting/Gem/PythonTests/Prefab/TestSuite_Periodic.py [hunk 1]
```diff
@@ -18,5 +18,15 @@ class TestAutomationNoAutoTestMode(EditorTestSuite):
...
-        from .tests.delete_prefab import DeletePrefab_DuplicatedPrefabInstance as test_module
\ No newline at end of file
+    class test_AddEntity_UnderLevelPrefab(EditorSharedTest):
+        from .tests.add_entity import AddEntity_UnderLevelPrefab as test_module
+
+    class test_AddEntity_UnderAnotherEntity(EditorSharedTest):
+       from .tests.add_entity import AddEntity_UnderAnotherEntity as test_module
+
+    class test_AddEntity_UnderAPrefab(EditorSharedTest):
+        from .tests.add_entity import AddEntity_UnderAPrefab as test_module
```
Comment (forkercat) [inline AutomatedTesting/Gem/PythonTests/Prefab/TestSuite_Periodic.py:28]: minor suggest: test_AddEntity_UnderPrefab

Comment (dm-amzn) [reply AutomatedTesting/Gem/PythonTests/Prefab/TestSuite_Periodic.py]: Re-added missing test data and updated test name with feedback.

Comment (dm-amzn) [reply AutomatedTesting/Gem/PythonTests/Prefab/TestSuite_Periodic.py]: Re-added missing test data and updated test name with feedback.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderAnotherEntity.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    # Creates a new Entity at the root level
+    # Asserts if creation didn't succeed
+    parent_entity = EditorEntity.create_editor_entity_at((100.0, 100.0, 100.0))
+    assert parent_entity.id.IsValid(), "Couldn't create parent entity"
+
+    # Creates a new child Entity
+    # Asserts if creation didn't succeed
+    child_entity = EditorEntity.create_editor_entity(parent_id=parent_entity.id)
+    assert child_entity.id.IsValid(), "Couldn't create child entity"
+
```
Comment (forkercat) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderAnotherEntity.py:31]: apart from this validation, do we need to verify the relation?
```python
assert parent_entity.get_children_ids()[0] == child_entity.id, "..."
```

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderAnotherEntity.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+For complete copyright and license terms please see the LICENSE at the root of this distribution.
+
+SPDX-License-Identifier: Apache-2.0 OR MIT
+"""
+
+def AddEntity_UnderAnotherEntity():
+    """
+    Test description:
+    - Creates two entities, parent and child. Child entity has Parent entity as its parent.
+    - Creates a prefab of the child entity.
```
Comment (forkercat) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderAnotherEntity.py:12]: nit: typo?

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                                              f" when there should have been just 1 entity"
+
+    child_entity_inside_prefab = child_entity_ids_inside_prefab[0]
+    child_entity_inside_prefab.focus_on_owning_prefab()
+
+    second_child_entity = child_entity_inside_prefab.create_editor_entity(parent_id=child_entity_inside_prefab.id)
+
+    # Wait till prefab propagation finishes before validating add child entity
+    PrefabWaiter.wait_for_propagation()
+
```
Comment (forkercat) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py:46]: Is there a validation step missing here?

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    assert len(child_entity_ids_inside_prefab) == 0, f"{len(child_entity_ids_inside_prefab)} entities found inside prefab" \
+                                              f" after Undo operation, when there should have been 0 entity"
+    azlmbr.legacy.general.redo()
+    PrefabWaiter.wait_for_propagation()
+    child_entity_ids_inside_prefab = child_entity_inside_prefab.get_children()
+    assert len(child_entity_ids_inside_prefab) == 1, f"{len(child_entity_ids_inside_prefab)} entities found inside prefab" \
+                                              f" after Redo operation, when there should have been 1 entity"
+
+
+
```
Comment (forkercat) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py:61]: nit: extra empty line

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderPrefab.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+SPDX-License-Identifier: Apache-2.0 OR MIT
+"""
+
+def AddEntity_UnderPrefab():
+    """
+    Test description:
+    - Creates an entity.
+    - Creates a prefab out of the above entity.
+    - Focuses on the created prefab and adds a new child entity.
+    Checks that the entity is correctly added.
```
Comment (forkercat) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderPrefab.py:14]: nit: missing "-"

Comment (dm-amzn) [reply AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderPrefab.py]: Addressed feedback with those 3 tests in latest commit.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    from editor_python_test_tools.editor_entity_utils import EditorEntity
+    from editor_python_test_tools.prefab_utils import Prefab
+    from editor_python_test_tools.wait_utils import PrefabWaiter
+
+    import Prefab.tests.PrefabTestUtils as prefab_test_utils
+
+    prefab_test_utils.open_base_tests_level()
+
+    PREFAB_FILE_NAME = 'some_prefab'
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py:26]: Suggest using a more distinctive file name for this, as if tests are performed in batch mode, this test's prefab created in memory may collide with that set in AddEntity_UnderPrefab.py that uses the same filename. Previous tests have used the path to the test file + a suffix to create a unique in-memory prefab name.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    second_child_entity = child_entity_inside_prefab.create_editor_entity(parent_id=child_entity_inside_prefab.id)
+
+    # Wait till prefab propagation finishes before validating add child entity
+    PrefabWaiter.wait_for_propagation()
+    assert second_child_entity.id.IsValid(), "Couldn't create second child entity"
+
+    # Test undo/redo on add child entity
+    azlmbr.legacy.general.undo()
+    PrefabWaiter.wait_for_propagation()
+    child_entity_ids_inside_prefab = child_entity_inside_prefab.get_children()
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py:51]: This variable name is confusing when compared with line 35. There we checked for direct child entities of the instance as a whole. Here we're just checking a specific entity inside the instance for children. Suggest renaming this variable to be more clear about what it represents.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    assert len(child_entity_ids_inside_prefab) == 0, f"{len(child_entity_ids_inside_prefab)} entities found inside prefab" \
+                                              f" after Undo operation, when there should have been 0 entity"
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py:53]: This assert statement is inaccurate. There *should* be 1 entity inside the prefab, there should not, however, be a child of that entity inside the prefab.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    azlmbr.legacy.general.redo()
+    PrefabWaiter.wait_for_propagation()
+    child_entity_ids_inside_prefab = child_entity_inside_prefab.get_children()
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py:56]: Same issue as line 51.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    assert len(child_entity_ids_inside_prefab) == 1, f"{len(child_entity_ids_inside_prefab)} entities found inside prefab" \
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderChildEntityOfAnotherPrefab.py:57]: Same issue as lines 52-53, but with updated numbers on expectations. Here we expect there to be 2 entities inside the prefab, however we expect 1 of those to be a child entity.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderPrefab.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+    from editor_python_test_tools.editor_entity_utils import EditorEntity
+    from editor_python_test_tools.prefab_utils import Prefab
+    from editor_python_test_tools.wait_utils import PrefabWaiter
+
+    import Prefab.tests.PrefabTestUtils as prefab_test_utils
+
+    prefab_test_utils.open_base_tests_level()
+
+    PREFAB_FILE_NAME = 'some_prefab'
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderPrefab.py:25]: Prefab file name should be unique to this test.

File: AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderPrefab.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    # Test undo/redo on add child entity
+    azlmbr.legacy.general.undo()
+    PrefabWaiter.wait_for_propagation()
+    child_entity_ids_inside_prefab = child_instance.get_direct_child_entities()
+    assert len(child_entity_ids_inside_prefab) == 0, f"{len(child_entity_ids_inside_prefab)} entities found inside prefab" \
+                                              f" after Undo operation, when there should have been 0 entities"
+    azlmbr.legacy.general.redo()
+    PrefabWaiter.wait_for_propagation()
+    child_entity_ids_inside_prefab = child_instance.get_direct_child_entities()
+    assert len(child_entity_ids_inside_prefab) == 1, f"{len(child_entity_ids_inside_prefab)} entities found inside prefab" \
```
Comment (jckand-amzn) [inline AutomatedTesting/Gem/PythonTests/Prefab/tests/add_entity/AddEntity_UnderPrefab.py:47]: This all seems inaccurate. You're creating a prefab from a single entity, and then undoing that change. That should just undo the creation of the prefab, leaving only a bare entity behind with no prefab at all. When you redo, you're recreating the prefab, so this test is really just verifying that creating a prefab and undo/redo work which is already covered in CreatePrefab_WithSingleEntity.py.

**Review by jckand-amzn — CHANGES_REQUESTED** [review]

**Review by jckand-amzn — APPROVED** [review]

**Review by forkercat — APPROVED** [review]

**Pull request closed without merging**